src/ape_ethereum/ecosystem.py

```diff
@@ -320,11 +325,69 @@ def create_transaction(self, **kwargs) -> TransactionAPI:
 
         return txn_class(**kwargs)  # type: ignore
 
-    def decode_event(self, abi: EventABI, receipt: "ReceiptAPI") -> "ContractLog":
-        filter_id = keccak(to_bytes(text=abi.selector))
-        event_data = next(log for log in receipt.logs if log["filter_id"] == filter_id)
+    def decode_logs(self, abi: EventABI, data: List[Dict]) -> Iterator["ContractLog"]:
+        event_id_bytes = keccak(to_bytes(text=abi.selector))
+        matching_logs = [log for log in data if log["topics"][0] == event_id_bytes]
+        if not matching_logs:
+            raise DecodingError(f"No logs found with ID '{abi.selector}'.")
+
+        # Process indexed data (topics)
+        log_topics = [i for i in abi.inputs if i.indexed]
+        log_topic_names = [abi.name for abi in log_topics]
+        normalized_log_topics = [abi for abi in _normalize_abi_types(log_topics)]
+        log_topic_types = [t for t in _get_abi_types(normalized_log_topics)]
+
+        # Process the rest of the log data
+        log_data = [i for i in abi.inputs if not i.indexed]
+        log_data_names = [abi.name for abi in log_data]
+        normalized_log_data = [abi for abi in _normalize_abi_types(log_data)]
+        log_data_types = [t for t in _get_abi_types(normalized_log_data)]
+
+        # Sanity check that there are not name intersections between the topic
+        # names and the data argument names.
+        duplicate_names = set(log_topic_names).intersection(log_data_names)
+        if duplicate_names:
+            duplicate_names_str = ", ".join(duplicate_names)
+            raise DecodingError(
+                "The following argument names are duplicated "
+                f"between event inputs: '{duplicate_names_str}'."
+            )
+
+        for log in matching_logs:
+            indexed_data = log["topics"] if log.get("anonymous", False) else log["topics"][1:]
+            log_data = hexstr_if_str(to_bytes, log["data"])
+
+            if len(indexed_data) != len(log_topic_types):
+                raise DecodingError(
+                    f"Expected '{len(indexed_data)}' log topics.  Got '{len(log_topic_types)}'."
+                )
+
+            decoded_topic_data = [
+                decode_single(topic_type, topic_data)
+                for topic_type, topic_data in zip(log_topic_types, indexed_data)
+            ]
+            decoded_log_data = decode_abi(log_data_types, log_data)
+            event_args = dict(
+                itertools.chain(
+                    zip(log_topic_names, decoded_topic_data),
+                    zip(log_data_names, decoded_log_data),
+                )
+            )
+            yield ContractLog(name=abi.name, data=event_args)  # type: ignore
+
+
+def _normalize_abi_types(abi_inputs: List[EventABIType]) -> Iterator[Dict]:
+    for abi_input in abi_inputs:
+        if re.match(r"^{lib_name}r'\.{enum_name}$", str(abi_input.type)):
```

Comment:
This is kind of wild, need to understand this more

Reply:
This is wrong (and not the greated, copied code), but it has to do with flattening enum types.

However, looking at `ethpm-types`, I think this may already happen via `canonical_types` but I will have to test it out